ee/query-service/anomaly/params.go

```diff
@@ -0,0 +1,191 @@
+			q.StepInterval = minStep
+		}
+	}
+}
+
+func prepareAnomalyQueryParams(req *v3.QueryRangeParamsV3, seasonality Seasonality) *anomalyQueryParams {
+	start := req.Start
+	end := req.End
+
+	currentPeriodQuery := &v3.QueryRangeParamsV3{
+		Start:          start,
+		End:            end,
+		Step:           int64(math.Max(float64(common.MinAllowedStepInterval(start, end)), 60)),
+		CompositeQuery: req.CompositeQuery,
+		Variables:      make(map[string]interface{}, 0),
+		NoCache:        false,
+	}
+	updateStepInterval(currentPeriodQuery)
+
+	var pastPeriodStart, pastPeriodEnd int64
+
+	switch seasonality {
+	// for one week period, we fetch the data from the past week with 4 hours offset
+	case SeasonalityWeekly:
+		pastPeriodStart = start - 166*time.Hour.Milliseconds() - 4*time.Hour.Milliseconds()
+		pastPeriodEnd = end - 166*time.Hour.Milliseconds()
+	// for one day period, we fetch the data from the past day with 2 hours offset
+	case SeasonalityDaily:
+		pastPeriodStart = start - 23*time.Hour.Milliseconds() - 2*time.Hour.Milliseconds()
+		pastPeriodEnd = end - 23*time.Hour.Milliseconds()
+	// for one hour period, we fetch the data from the past hour with 30 minutes offset
+	case SeasonalityHourly:
+		pastPeriodStart = start - 1*time.Hour.Milliseconds() - 30*time.Minute.Milliseconds()
+		pastPeriodEnd = end - 1*time.Hour.Milliseconds()
+	}
+
+	pastPeriodQuery := &v3.QueryRangeParamsV3{
+		Start:          pastPeriodStart,
+		End:            pastPeriodEnd,
+		Step:           int64(math.Max(float64(common.MinAllowedStepInterval(pastPeriodStart, pastPeriodEnd)), 60)),
+		CompositeQuery: copyCompositeQuery(req),
+		Variables:      make(map[string]interface{}, 0),
+		NoCache:        false,
+	}
+	updateStepInterval(pastPeriodQuery)
+
+	// seasonality growth trend
+	var currentGrowthPeriodStart, currentGrowthPeriodEnd int64
+	switch seasonality {
+	case SeasonalityWeekly:
+		currentGrowthPeriodStart = start - 7*24*time.Hour.Milliseconds()
+		currentGrowthPeriodEnd = end
+	case SeasonalityDaily:
+		currentGrowthPeriodStart = start - 23*time.Hour.Milliseconds()
+		currentGrowthPeriodEnd = end
+	case SeasonalityHourly:
+		currentGrowthPeriodStart = start - 1*time.Hour.Milliseconds()
+		currentGrowthPeriodEnd = end
+	}
+
+	currentGrowthQuery := &v3.QueryRangeParamsV3{
+		Start:          currentGrowthPeriodStart,
+		End:            currentGrowthPeriodEnd,
+		Step:           int64(math.Max(float64(common.MinAllowedStepInterval(currentGrowthPeriodStart, currentGrowthPeriodEnd)), 60)),
+		CompositeQuery: copyCompositeQuery(req),
+		Variables:      make(map[string]interface{}, 0),
+		NoCache:        false,
+	}
+	updateStepInterval(currentGrowthQuery)
+
+	var pastGrowthPeriodStart, pastGrowthPeriodEnd int64
+	switch seasonality {
+	case SeasonalityWeekly:
+		pastGrowthPeriodStart = start - 14*24*time.Hour.Milliseconds()
+		pastGrowthPeriodEnd = start - 7*24*time.Hour.Milliseconds()
```

Comment:
nit : pastGrowthPeriodEnd = start - currentGrowthPeriodStart will be better?

Reply:
probably but i want to keep it explicit in terms of how much we offset.